src/identity/webauthn/webauthn-identity.ts

```diff
@@ -0,0 +1,151 @@
+import cbor from "cbor"
+import { CoseKey, EMPTY } from "../../message/cose"
+import { Identity } from "../types"
+const sha512 = require("js-sha512")
+
+const CHALLENGE_BUFFER = new TextEncoder().encode("lifted")
+const ONE_MINUTE = 60000
+
+export class WebAuthnIdentity extends Identity {
+  publicKey: ArrayBuffer
+  rawId: ArrayBuffer
+  cosePublicKey: ArrayBuffer
+
+  constructor(cosePublicKey: ArrayBuffer, rawId: ArrayBuffer) {
+    super()
+    this.cosePublicKey = cosePublicKey
+    this.publicKey = this.getPublicKeyFromCoseKey(cosePublicKey)
+    this.rawId = rawId
+  }
+
+  private getPublicKeyFromCoseKey(cosePublicKey: ArrayBuffer): ArrayBuffer {
+    const decoded = cbor.decodeFirstSync(cosePublicKey)
+    return decoded.get(-2)
+  }
+
+  static async create(): Promise<WebAuthnIdentity> {
+    const publicKeyCredential = await createPublicKeyCredential()
+    const attestationResponse = publicKeyCredential?.response
+    if (!(attestationResponse instanceof AuthenticatorAttestationResponse)) {
+      throw new Error("Must be AuthenticatorAttestationResponse")
+    }
+    const attestationObj = cbor.decodeFirstSync(
+      attestationResponse.attestationObject,
+    )
+    const cosePublicKey = getCosePublicKey(attestationObj.authData)
+    return new WebAuthnIdentity(cosePublicKey, publicKeyCredential.rawId)
+  }
+
+  async sign(): Promise<ArrayBuffer> {
+    return EMPTY
+  }
+
+  async verify(_: ArrayBuffer): Promise<boolean> {
+    return false
+  }
+
+  static async getCredential(
+    credentialId: ArrayBuffer,
+    challenge?: Uint8Array | ArrayBuffer,
+  ): Promise<PublicKeyCredential> {
+    let credential = (await window.navigator.credentials.get({
+      publicKey: {
+        challenge: challenge ?? CHALLENGE_BUFFER,
+        timeout: ONE_MINUTE,
+        userVerification: "discouraged",
+        allowCredentials: [
+          {
+            transports: ["nfc", "usb"],
+            id: credentialId,
+            type: "public-key",
+          },
+        ],
+      },
+    })) as PublicKeyCredential
+    return credential
+  }
+
+  async getUnprotectedHeader(
+    cborMessageContent: ArrayBuffer,
+    cborProtectedHeader: ArrayBuffer,
+  ): Promise<Map<string, unknown>> {
+    const c = new Map()
+    c.set(0, cborProtectedHeader)
+    c.set(
+      1,
+      Buffer.from(sha512.arrayBuffer(cborMessageContent)).toString("base64"),
+    )
+    const challenge = cbor.encode(c)
+    const cred = await WebAuthnIdentity.getCredential(this.rawId, challenge)
+    const response = cred.response as AuthenticatorAssertionResponse
+    const m = new Map()
+    m.set("webauthn", true)
+    m.set("authData", response.authenticatorData)
+    m.set("clientData", Buffer.from(response.clientDataJSON).toString())
+    m.set("signature", response.signature)
```

Comment:
@fmorency this line adds `signature` in the unprotected headers. I think you use the regular signature field now, right? So this would be optional.

Reply:
@hansl No, the regular signature field contains an empty array at the moment. 

We could probably move the signature from the unprotected header to the standard field unless I'm missing something.